### PR TITLE
Skeleton main runtime

### PR DESCRIPTION
**Pull request by carlbraganza**
Initial pass at the main program setup using the `external-snapshotter` repo's `csi-snapshotter` main as a template.

The PR attempts to keep `main.go` very thin by moving most of the runtime setup to `pkg/runtime`.
Current validation logic is incomplete because a CSI release containing the `SnapshotMetadata` service is not available.

### CODE REVIEW

**Comment by carlbraganza** [comment]
There is no top-level `Makefile` currently - need to pull in the release tools etc - but one can build this from the top of the repo as follows:
```
$ (cd cmd/csi-snapshot-metadata-sidecar/; go build -ldflags '-X main.version="test-build"')
```
The sidecar can be executed as follows:
```
# see version
$ cmd/csi-snapshot-metadata-sidecar/csi-snapshot-metadata-sidecar -version
cmd/csi-snapshot-metadata-sidecar/csi-snapshot-metadata-sidecar "test-build"

# see usage
$ cmd/csi-snapshot-metadata-sidecar/csi-snapshot-metadata-sidecar -h
Usage of cmd/csi-snapshot-metadata-sidecar/csi-snapshot-metadata-sidecar:
 ...
 
 # launch
 $ cmd/csi-snapshot-metadata-sidecar/csi-snapshot-metadata-sidecar
I0617 19:11:47.270991  111478 sidecar.go:45] Version: "test-build"
E0617 19:11:47.271056  111478 sidecar.go:57] error in cluster configuration: unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined
```

File: pkg/sidecar/sidecar.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	})
+	if err != nil {
+		klog.Error(err)
+		return 1
+	}
+
+	klog.Infof("CSI driver name: %q", rt.DriverName)
+
+	// TBD
+	// May need to exposed metric and healthz HTTP end points
```
Comment (PrasadG193) [inline pkg/sidecar/sidecar.go:64]: For the server health endpoint, we can use https://pkg.go.dev/google.golang.org/grpc/health and register health service on the same port. For metrics, as discussed it makes sense to start a new HTTP service on a different port.

**Comment by carlbraganza** [comment]
/remove-approved
Not sure how the bot thinks I approved this!

**Comment by carlbraganza** [comment]
Rebased on latest main

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	github.com/kubernetes-csi/csi-test/v5 v5.2.0
+	google.golang.org/grpc v1.64.0
+	k8s.io/client-go v0.30.0
+	k8s.io/klog/v2 v2.120.1
+)
+
+require (
+	github.com/stretchr/testify v1.8.4
+	google.golang.org/protobuf v1.34.1
+)
```
Comment (PrasadG193) [inline go.mod:19]:
```suggestion
	k8s.io/klog/v2 v2.120.1
	github.com/stretchr/testify v1.8.4
	google.golang.org/protobuf v1.34.1
)
```
can we club direct dependencies in a single group?

File: pkg/internal/runtime/runtime.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	"fmt"
+	"time"
+
+	"github.com/kubernetes-csi/csi-lib-utils/connection"
+	"github.com/kubernetes-csi/csi-lib-utils/metrics"
+	csirpc "github.com/kubernetes-csi/csi-lib-utils/rpc"
+
+	"google.golang.org/grpc"
+
+	"k8s.io/client-go/kubernetes"
```
Comment (PrasadG193) [inline pkg/internal/runtime/runtime.go:31]:
```suggestion
	csirpc "github.com/kubernetes-csi/csi-lib-utils/rpc"
	"google.golang.org/grpc"
	"k8s.io/client-go/kubernetes"
```
All the imports are 3rd-party, so can be put in a same group

**Review by PrasadG193 — APPROVED** [review]
Minor suggestions on grouping imports. Overall LGTM

**Comment by PrasadG193** [comment]
@carlbraganza could you please squash the commits into a single commit? The bot doesn't do it while merging it

**Comment by carlbraganza** [comment]
Squashed commits.

**Comment by carlbraganza** [comment]
/release-note-edit
```
Initial pass at the main program setup using the external-snapshotter repo's csi-snapshotter main as a template.
The `main.c` program is kept very thin and most logic is in the `pkg` branch to facilitate unit testing.
```

**Comment by Rakshith-R** [comment]
/release-note-edit Initial pass at the main program setup using the external-snapshotter repo's csi-snapshotter main as a template. The `main.c` program is kept very thin and most logic is in the `pkg` branch to facilitate unit testing.

**Comment by Rakshith-R** [comment]
/release-note-edit 
```release-note
Initial pass at the main program setup using the external-snapshotter repo's csi-snapshotter main as a template. The main.c program is kept very thin and most logic is in the pkg branch to facilitate unit testing.
```

**Comment by Rakshith-R** [comment]
/release-note-edit ```release-note
Initial pass at the main program setup using the external-snapshotter repo's csi-snapshotter main as a template. The main.c program is kept very thin and most logic is in the pkg branch to facilitate unit testing.
```

**Comment by Rakshith-R** [comment]
/release-note-edit 
```release-note
Add skeleton main runtime.
```

**Pull request merged**